internal/controller/progressive/progressive.go

```diff
@@ -296,3 +329,124 @@ func IsNumaflowChildReady(upgradingObjectStatus *kubernetes.GenericStatus) bool
 	}
 	return true
 }
+
+/*
+scaleDownPromotedChild attempts to scale down the source vertices of a promoted child.
+// It checks if scaling is required based on the rollout
+status and performs the scaling operation if necessary. The function updates the rollout
+status with the new scale values and returns whether scaling was performed.
+
+Parameters:
+  - ctx: The context for managing request-scoped values, cancellation, and timeouts.
+  - rolloutObject: The rollout object containing the current rollout status.
+  - liveRolloutObject: The live rollout object representing the current state.
+  - existingPromotedChild: The unstructured object of the promoted child to be scaled down.
+  - skipPromotedChildSourceVerticesScaling: A flag to skip scaling if set to true.
+  - controller: The progressive controller handling the scaling logic.
+  - c: The Kubernetes client for interacting with the cluster.
+
+Returns:
+  - A boolean indicating if scaling was performed.
+  - An error if any issues occur during the scaling process.
+*/
+func scaleDownPromotedChild(
+	ctx context.Context,
+	rolloutObject ctlrcommon.RolloutObject,
+	liveRolloutObject ctlrcommon.RolloutObject,
+	existingPromotedChild *unstructured.Unstructured,
+	skipPromotedChildSourceVerticesScaling bool,
+	controller progressiveController,
+	c client.Client,
+) (bool, error) {
+
+	if skipPromotedChildSourceVerticesScaling {
+		return false, nil
+	}
+
+	if !liveRolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus.AreAllSourceVerticesScaledDown(existingPromotedChild.GetName()) {
+
+		// ScaleDownPromotedChildSourceVertices either updates the existingPromotedChild to scale down the source vertices pods or
+		// retrieves the currently running pods to update the scaleValuesMap used on the rollout status.
+		// This serves to make sure that the pods for each vertex have been really scaled down before proceeding with the progressive update.
+		scaleValuesMap, promotedChildNeedsUpdate, err := controller.ScaleDownPromotedChildSourceVertices(ctx, liveRolloutObject, existingPromotedChild, c)
+		if err != nil {
+			return false, fmt.Errorf("error updating scaling properties to the existing promoted child definition: %w", err)
+		}
+
+		if promotedChildNeedsUpdate {
+			if err = kubernetes.UpdateResource(ctx, c, existingPromotedChild); err != nil {
```

Comment:
this can be a future PR, but we should consider making this a patch instead of an update. Then we won't have as much change of resource version conflict. We ended up changing the pause logic to do that FYI: https://github.com/numaproj/numaplane/blob/main/internal/controller/ppnd/ppnd_pause.go#L132-L136